src/compiler/checker.ts

```diff
@@ -25141,6 +25160,39 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             let inferredType: Type | undefined;
             const signature = context.signature;
             if (signature) {
+                if (inference.indexes) {

```

Comment:
At the time of this comment this aggregated inference is pushed as an extra candidate (that's the behavior preserved from https://github.com/microsoft/TypeScript/pull/20126 ). 

That creates problems with cases like this:
```ts
interface Expression extends Node {
    _expressionBrand: any;
}

declare function setParent<T extends Node>(child: T, parent: T["parent"] | undefined): T;

interface Node {
    readonly kind: number;
    readonly parent: Node;
}

declare const expr: Expression
declare const node: Node

const res = setParent(expr, node)
//    ^? actual: Node, expected: Expression

export {}
```

It incorrectly widens the inferred type because it's able to eliminate the inferred subtype (`Expression`) through subtype reduction~.

I was afraid of cases like this in my initial implementation and in there I decided to just use the gathered `indexes`~ when there were no candidates for the "naked" type variable. Should I do that here and avoid making the intersected `indexes` the "same priority" candidate?

Reply:
To clarify, this is what happens right now:
1. we end up with 2 candidates (because the `aggregateInference` is appended as an extra candidate): `Expression` and `{ parent: Node; }`
2. we compute a common super type (`getCommonSupertype(baseCandidates)`) which is `{ parent: Node; }`
3. that is not assignable to the constraint, it fails the `compareTypes` check and thus we assign the instantiated constraint as the inferred type (`inference.inferredType = inferredType = instantiatedConstraint`)
4. that in turn is just OK further down the road since the constraint (`Node`) satisfies everything here

I tried to just always require the `aggregateInference` to be comparable to the constraint but that broke the "discriminating logic" added for this `aggregateInference` stuff. So I reverted that change and only added this requirement for non-union constraints. 

At this point, the logic related to pushing `aggregateInference` into the `inference.candidates` is a little bit hairy 🤷‍♂️ I guess the main question here is if every bit of the included enhancement is good at the end of the day, how it fits the rest of the compiler etc. That's something on what you'd have to chime in with @ahejlsberg .

I still have some of [those change](https://github.com/microsoft/TypeScript/pull/53017/commits/d0e8b8b4b9e49447921d7c1646c92488bccb66bf) to investigate but at least the self-check is working now so perhaps we could create a build using this PR for toying purposes.